{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#done 23 Aug\n",
    "import os\n",
    "os.environ[\"CUDA_VISIBLE_DEVICES\"] = \"3\"\n",
    "import json\n",
    "import openai\n",
    "from dotenv import load_dotenv\n",
    "import matplotlib.pyplot as plt\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import pickle\n",
    "from datetime import datetime, timedelta\n",
    "from langchain.chat_models import ChatOpenAI\n",
    "from langchain.schema import HumanMessage, AIMessage, ChatMessage\n",
    "import scvi\n",
    "from matplotlib.pyplot import rc_context\n",
    "sc.set_figure_params(dpi=100)\n",
    "scvi._settings.seed = 0\n",
    "from scvi.model import SCVI\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "import ast\n",
    "import matplotlib\n",
    "import warnings\n",
    "import numpy as np\n",
    "# Suppress all warnings\n",
    "warnings.filterwarnings(\"ignore\", category=UserWarning)\n",
    "warnings.filterwarnings(\"ignore\", category=DeprecationWarning)\n",
    "\n",
    "openai.api_key = 'sk-xxx'\n",
    "openai.base_url= \"xxx/v1/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Founction tools of scChat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "current_adata = None\n",
    "base_annotated_adata = None\n",
    "adata = None\n",
    "resolution = 0.5\n",
    "sample_mapping = None\n",
    "SGP = None\n",
    "global function_flag\n",
    "function_flag = False\n",
    "global display_flag\n",
    "display_flag = False\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_and_load_sample_mapping(directory):\n",
    "    global sample_mapping\n",
    "\n",
    "    for root, dirs, files in os.walk(directory):\n",
    "        for file in files:\n",
    "            if file == 'sample_mapping.json':\n",
    "                file_path = os.path.join(root, file)\n",
    "                \n",
    "                with open(file_path, 'r') as f:\n",
    "                    sample_mapping = json.load(f)\n",
    "                \n",
    "                print(f\"'sample_mapping.json' found and loaded from {file_path}\")\n",
    "                return sample_mapping\n",
    "    \n",
    "    # If the file wasn't found\n",
    "    return None\n",
    "\n",
    "#tag True for RAG, False for marker genes list\n",
    "def get_rag_and_markers(tag, specification):\n",
    "    # specification = None\n",
    "    # # file_path = \"../media/specification.json\"\n",
    "    # file_path = \"media/specification.json\"\n",
    "    # if os.path.exists(file_path):\n",
    "    #     with open(file_path, 'r') as file:\n",
    "    #         specification = json.load(file)\n",
    "    #         # print(\"BASE:\", specification)\n",
    "    # else:\n",
    "    #     print (\"specification not found\")\n",
    "    #     return \"-\"\n",
    "\n",
    "    base_file_path = os.path.join(\"schatbot/scChat_RAG\", specification['marker'].lower())\n",
    "    file_paths = []\n",
    "    \n",
    "    for tissue in specification['tissue']:\n",
    "        file_path = os.path.join(base_file_path, tissue.lower(), specification['condition'] + '.json')\n",
    "        file_paths.append(file_path)\n",
    "    \n",
    "    print(\"Constructed file paths:\", file_paths)\n",
    "\n",
    "    for file_path in file_paths:\n",
    "        if os.path.exists(file_path):\n",
    "            print(f\"File found: {file_path}\")\n",
    "            with open(file_path, 'r') as file:\n",
    "                data = json.load(file)\n",
    "                # print(data)\n",
    "        else:\n",
    "            print(f\"File not found: {file_path}\")\n",
    "            continue\n",
    "    \n",
    "    combined_data = {}\n",
    "\n",
    "    # Iterate through the file paths\n",
    "    for file_path in file_paths:\n",
    "        if os.path.exists(file_path):\n",
    "            print(f\"File found: {file_path}\")\n",
    "            with open(file_path, 'r') as file:\n",
    "                data = json.load(file)\n",
    "\n",
    "                if tag:  # If tag is true, combine all data from the files\n",
    "                    for cell_type, cell_data in data.items():\n",
    "                        if cell_type not in combined_data:\n",
    "                            combined_data[cell_type] = cell_data\n",
    "                        else:\n",
    "                            combined_data[cell_type]['markers'].extend(cell_data['markers'])\n",
    "\n",
    "                else:  # If tag is false, retrieve only marker name + list of genes\n",
    "                    for cell_type, cell_data in data.items():\n",
    "                        if cell_type not in combined_data:\n",
    "                            combined_data[cell_type] = {'genes': []}\n",
    "                        combined_data[cell_type]['genes'].extend([marker['gene'] for marker in cell_data['markers']])\n",
    "        \n",
    "        else:\n",
    "            print(f\"File not found: {file_path}\")\n",
    "\n",
    "    fptr = open(\"testop.txt\", \"w\")\n",
    "    fptr.write(json.dumps(combined_data, indent=4))\n",
    "    # print(\"Combined data:\", json.dumps(combined_data, indent=4))\n",
    "    return combined_data\n",
    "\n",
    "def filter_existing_genes(adata, gene_list):\n",
    "    existing_genes = [gene for gene in gene_list if gene in adata.raw.var_names]\n",
    "    return existing_genes\n",
    "\n",
    "def extract_top_genes_stats(adata, groupby='leiden', n_genes=25):\n",
    "    result = adata.uns['rank_genes_groups']\n",
    "    gene_names = result['names']\n",
    "    pvals = result['pvals']\n",
    "    pvals_adj = result['pvals_adj']\n",
    "    logfoldchanges = result['logfoldchanges']\n",
    "    \n",
    "    top_genes_stats = {group: {} for group in gene_names.dtype.names}\n",
    "    \n",
    "    for group in gene_names.dtype.names:\n",
    "        top_genes_stats[group]['gene'] = gene_names[group][:n_genes]\n",
    "        top_genes_stats[group]['pval'] = pvals[group][:n_genes]\n",
    "        top_genes_stats[group]['pval_adj'] = pvals_adj[group][:n_genes]\n",
    "        top_genes_stats[group]['logfoldchange'] = logfoldchanges[group][:n_genes]\n",
    "    \n",
    "    top_genes_stats_df = pd.concat({group: pd.DataFrame(top_genes_stats[group])\n",
    "                                    for group in top_genes_stats}, axis=0)\n",
    "\n",
    "    top_genes_stats_df = top_genes_stats_df.reset_index()\n",
    "    \n",
    "    top_genes_stats_df = top_genes_stats_df.rename(columns={'level_0': 'cluster', 'level_1': 'index'})\n",
    "    return top_genes_stats_df\n",
    "\n",
    "# Define the statistical extraction and API interaction functions\n",
    "def calculate_cluster_statistics(adata, category, n_genes=25, specification=None):\n",
    "    #adding\n",
    "    global sample_mapping\n",
    "    base_markers = get_rag_and_markers(False,specification)\n",
    "    markers = []\n",
    "    for cell_type, cell_data in base_markers.items():\n",
    "        # print (cell_type)\n",
    "        # print ('--')\n",
    "        # print (cell_data)\n",
    "        markers += cell_data['genes']\n",
    "    \n",
    "    # print (\"MARKERS BEFORE FILTER2 \", markers)\n",
    "    markers = filter_existing_genes(adata, markers)\n",
    "    # print (\"MARKERS FINAL2 \", markers)\n",
    "    markers = list(set(markers))\n",
    "\n",
    "    sc.tl.rank_genes_groups(adata, 'leiden', method='wilcoxon', n_genes=n_genes)\n",
    "    top_genes_df = extract_top_genes_stats(adata, groupby='leiden', n_genes=25)\n",
    "    second_dataset = True\n",
    "\n",
    "    if sample_mapping:\n",
    "        sc.tl.dendrogram(adata, groupby='leiden', use_rep='X_scVI')\n",
    "        # sc.tl.dendrogram(adata, groupby='leiden')\n",
    "    else:\n",
    "        sc.tl.dendrogram(adata, groupby='leiden')\n",
    "\n",
    "    marker_expression = sc.get.obs_df(adata, keys=['leiden'] + markers, use_raw=True)\n",
    "    marker_expression.set_index('leiden', inplace=True)\n",
    "    # Calculating mean and proportion of expression per cluster\n",
    "    mean_expression = marker_expression.groupby('leiden').mean()\n",
    "    expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n",
    "    global global_top_genes_df, global_mean_expression, global_expression_proportion\n",
    "    global_top_genes_df = top_genes_df\n",
    "    global_mean_expression = mean_expression\n",
    "    global_expression_proportion = expression_proportion\n",
    "    return global_top_genes_df, global_mean_expression, global_expression_proportion\n",
    "\n",
    "def retreive_stats(specification):\n",
    "    with open(\"basic_data/mean_expression.json\", 'r') as file:\n",
    "        mean_expression = json.load(file)\n",
    "    with open(\"basic_data/expression_proportion.json\", 'r') as file:\n",
    "        expression_proportion = json.load(file)\n",
    "    global adata\n",
    "    global_top_genes_df, global_mean_expression, global_expression_proportion = calculate_cluster_statistics(adata, 'overall', specification=specification)\n",
    "    # myeloid_markers = get_myeloid_markers()\n",
    "    # t_cell_markers = get_t_markers()\n",
    "    # overall_markers = get_overall_markers()    \n",
    "    markers = get_rag_and_markers(False, specification)\n",
    "    markers = ', '.join(markers)\n",
    "    explanation = \"Please analyze the clustering statistics and classify each cluster based on the following data: Top Genes:Mean Expression: Expression Proportion: , based on statistical data: 1. top_genes_df: 25 top genes expression within each clusters, with it's p_val, p_val_adj, and logfoldchange; 2. mean_expression of the marker genes: specific marker genes mean expression within each cluster; 3. expression_proportion of the marker genes: every cluster each gene expression fraction within each cluster, and give back the mapping dictionary in the format like this group_to_cell_type = {'0': 'Myeloid cells','1': 'T cells','2': 'Myeloid cells','3': 'Myeloid cells','4': 'T cells'} without further explanation or comment.  I only want the summary map in the response, do not give me any explanation or comment or repeat my input, i dont want any redundant information other than the summary map\"\n",
    "    top_genes_summary = []\n",
    "    mean_expression_str = \", \".join([f\"{k}: {v}\" for k, v in mean_expression.items()])\n",
    "    expression_proportion_str = \", \".join([f\"{k}: {v}\" for k, v in expression_proportion.items()])\n",
    "    # myeloid_markers_str = \", \".join(myeloid_markers)\n",
    "    # t_cell_markers_str = \", \".join(t_cell_markers)\n",
    "    # overall_markers_str = \", \".join(overall_markers)\n",
    "    summary = (\n",
    "        f\"Explanation: {explanation}. \"\n",
    "        f\"Mean expression data: {mean_expression_str}. \"\n",
    "        f\"Expression proportion data: {expression_proportion_str}. \"\n",
    "        f\"Top genes details: {global_top_genes_df}. \"\n",
    "        f\"markers: {markers}. \"\n",
    "    )\n",
    "    return summary\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "globalGroupToCellType = None\n",
    "def generate_umap(path_row):\n",
    "    global sample_mapping\n",
    "    second_dataset = True\n",
    "    matplotlib.use('Agg')\n",
    "    global adata\n",
    "    global resolution\n",
    "    path = path_row #find_file_with_extension(\"media\", \".h5ad\")\n",
    "    if not path:\n",
    "        return \".h5ad file isn't given, unable to generate UMAP.\"\n",
    "    adata = sc.read_h5ad(path)\n",
    "    organ = path.split('/')[-1].split('.')[0]\n",
    "\n",
    "    global current_adata\n",
    "    current_adata = adata\n",
    "\n",
    "    # Data preprocessing\n",
    "    sc.pp.filter_cells(adata, min_genes=100)\n",
    "    sc.pp.filter_genes(adata, min_cells=3)\n",
    "    adata.var['mt'] = adata.var_names.str.startswith('MT-')\n",
    "    sc.pp.calculate_qc_metrics(adata, qc_vars=['mt'], percent_top=None, log1p=False, inplace=True)\n",
    "    adata = adata[adata.obs.pct_counts_mt < 20]\n",
    "    adata.layers['counts'] = adata.X.copy()  # used by scVI-tools\n",
    "\n",
    "    # Normalization\n",
    "    sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "    sc.pp.log1p(adata)\n",
    "    adata.raw = adata\n",
    "\n",
    "    # find_and_load_sample_mapping(\"media\")\n",
    "    # Variable genes\n",
    "    if sample_mapping:\n",
    "        sc.pp.highly_variable_genes(adata, n_top_genes=3000, subset=True, layer='counts', flavor=\"seurat_v3\", batch_key=\"Sample\")\n",
    "    else:\n",
    "        sc.pp.highly_variable_genes(adata, n_top_genes=3000, subset=True, layer='counts', flavor=\"seurat_v3\")\n",
    "\n",
    "    # Setup and load scVI model\n",
    "    if sample_mapping:\n",
    "        SCVI.setup_anndata(adata, layer=\"counts\", categorical_covariate_keys=[\"Sample\"], continuous_covariate_keys=['pct_counts_mt', 'total_counts'])\n",
    "        model = SCVI.load(dir_path=\"schatbot/glioma_scvi_model\", adata=adata)\n",
    "        latent = model.get_latent_representation()\n",
    "        adata.obsm['X_scVI'] = latent\n",
    "        adata.layers['scvi_normalized'] = model.get_normalized_expression(library_size=1e4)\n",
    "    \n",
    "    # Clustering and UMAP\n",
    "    if sample_mapping:\n",
    "        sc.pp.neighbors(adata, use_rep='X_scVI')\n",
    "    else:\n",
    "        sc.pp.neighbors(adata)\n",
    "    sc.tl.umap(adata)\n",
    "    sc.tl.leiden(adata, resolution=resolution)\n",
    "\n",
    "\n",
    "    if sample_mapping:\n",
    "        adata.obs['patient_name'] = adata.obs['Sample'].map(sample_mapping)\n",
    "\n",
    "    # UMAP plot\n",
    "    umap_df = adata.obsm['X_umap']\n",
    "    adata.obs['UMAP_1'] = umap_df[:, 0]\n",
    "    adata.obs['UMAP_2'] = umap_df[:, 1]\n",
    "\n",
    "    #take this from MARK\n",
    "    # markers = get_overall_markers()\n",
    "    specification = {'marker':'human','tissue':[organ.lower()],'condition':'normal'}\n",
    "    base_markers = get_rag_and_markers(False, specification)\n",
    "    markers = []\n",
    "    for cell_type, cell_data in base_markers.items():\n",
    "        print (cell_type)\n",
    "        print ('--')\n",
    "        print (cell_data)\n",
    "        markers += cell_data['genes']\n",
    "    \n",
    "    # print (\"MARKERS BEFORE FILTER \", markers)\n",
    "    markers = filter_existing_genes(adata, markers)\n",
    "    # print (\"MARKERS FINAL \", markers)\n",
    "    markers = list(set(markers))\n",
    "    # Calculate statistics to feed into GPT\n",
    "    statistic_data = sc.get.obs_df(adata, keys=['leiden'] + markers, use_raw=True)\n",
    "    statistic_data.set_index('leiden', inplace=True)\n",
    "    mean_expression = statistic_data.groupby('leiden').mean()\n",
    "    pd_mean_expression = pd.DataFrame(mean_expression)\n",
    "    pd_mean_expression.to_csv(\"basic_data/mean_expression.csv\")\n",
    "    pd_mean_expression.to_json(\"basic_data/mean_expression.json\")\n",
    "\n",
    "    expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
    "    pd_expression_proportion = pd.DataFrame(expression_proportion)\n",
    "    pd_expression_proportion.to_csv(\"basic_data/expression_proportion.csv\")\n",
    "    pd_expression_proportion.to_json(\"basic_data/expression_proportion.json\")\n",
    "    # Dot plot\n",
    "    if sample_mapping == None:\n",
    "        sc.tl.dendrogram(adata, groupby='leiden')\n",
    "        # with rc_context({'figure.figsize': (10, 10)}):\n",
    "        #     sc.pl.dotplot(adata, markers, groupby='leiden', swap_axes=True, use_raw=True, standard_scale='var', dendrogram=True, color_map=\"Blues\", save=\"dotplot.png\")\n",
    "        #     plt.close()\n",
    "    else:\n",
    "        sc.tl.dendrogram(adata, groupby='leiden', use_rep='X_scVI')\n",
    "        # with rc_context({'figure.figsize': (10, 10)}):\n",
    "        #     sc.pl.dotplot(adata, markers, groupby='leiden', swap_axes=True, use_raw=True, standard_scale='var', dendrogram=True, color_map=\"Blues\", save=\"dotplot.png\")\n",
    "        #     plt.close()\n",
    "    \n",
    "    adata.obs['cell_type'] = 'Unknown'\n",
    "    # adata.obs[['UMAP_1', 'UMAP_2', 'leiden', 'patient_name']].to_csv(\"basic_data/Overall cells_umap_data.csv\", index=False)\n",
    "    # adata.obs[['UMAP_1', 'UMAP_2', 'leiden', 'patient_name']].to_csv(\"process_cell_data/Overall cells_umap_data.csv\", index=False)\n",
    "\n",
    "    # Save dot plot data for Plotly\n",
    "    dot_plot_data = statistic_data.reset_index().melt(id_vars='leiden', var_name='gene', value_name='expression')\n",
    "    dot_plot_data.to_csv(\"basic_data/dot_plot_data.csv\", index=False)\n",
    "\n",
    "    # Save dendrogram data for Plotly\n",
    "    dendrogram_data = adata.uns['dendrogram_leiden']\n",
    "    pd_dendrogram_linkage = pd.DataFrame(dendrogram_data['linkage'], columns=['source', 'target', 'distance', 'count'])\n",
    "    pd_dendrogram_linkage.to_csv(\"basic_data/dendrogram_data.csv\", index=False)\n",
    "    # rag_data = load_RAG(\"Overall cells\")\n",
    "    rag_data = get_rag_and_markers(True, specification)\n",
    "    rag_data_str = ', '.join(rag_data)\n",
    "                #   f\"RAG Data : {', '.join(rag_data_str)}. \" \\\n",
    "\n",
    "    summary = f\"UMAP analysis completed. Data summary: {adata}, \" \\\n",
    "                f\"RAG Data : {str(rag_data)}. \" \\\n",
    "              f\"Cell counts details are provided. \" \\\n",
    "              f\"Additional data file generated: preface.txt.\"\n",
    "    retrieve_stats_summary = retreive_stats(specification)\n",
    "    final_summary = f\"{summary} {retrieve_stats_summary}\"\n",
    "    current_adata = adata\n",
    "\n",
    "    return final_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def label_clusters(cell_type, last_message):\n",
    "    global adata\n",
    "    global annotated_adata\n",
    "    global conversation_history2\n",
    "    global current_adata\n",
    "    global base_annotated_adata\n",
    "    adata2 = adata.copy()\n",
    "    standardized_cell_type3 = cell_type.split()[0].capitalize()     \n",
    "    standardized_cell_type2 = cell_type.split()[0].capitalize() + \" cell\"        \n",
    "    standardized_cell_type = cell_type.split()[0].capitalize() + \" cells\"        \n",
    "    # last_message = conversation_history2[-2]['content']\n",
    "    try:\n",
    "        start_idx = last_message.find(\"{\")\n",
    "        end_idx = last_message.rfind(\"}\") + 1\n",
    "        str_map = last_message[start_idx:end_idx]        \n",
    "        map2 = ast.literal_eval(str_map)\n",
    "        map2 = {str(key): value for key, value in map2.items()}        \n",
    "        if standardized_cell_type == \"Overall cells\" or standardized_cell_type2 == \"Overall cell\":\n",
    "            adata2.obs['cell_type_scChat'] = 'Unknown'\n",
    "            for group, cell_type in map2.items():\n",
    "                adata2.obs.loc[adata2.obs['leiden'] == group, 'cell_type_scChat'] = cell_type\n",
    "            # adata2.obs[['UMAP_1', 'UMAP_2', 'leiden', 'patient_name', 'cell_type_scChat']].to_csv(f'scChat/umaps/{standardized_cell_type}_annotated_umap_data.csv', index=False)\n",
    "            annotated_adata = adata2.copy()\n",
    "            fname = f'annotated_adata/Overall cells_annotated_adata.pkl'\n",
    "            with open(fname, \"wb\") as file:\n",
    "                pickle.dump(annotated_adata, file)\n",
    "            base_annotated_adata = adata2\n",
    "        else:\n",
    "            adata3 = base_annotated_adata.copy()\n",
    "            specific_cells = adata3[adata3.obs['cell_type_scChat'].isin([standardized_cell_type])].copy()\n",
    "            if specific_cells.shape[0] == 0:\n",
    "                specific_cells = adata3[adata3.obs['cell_type_scChat'].isin([standardized_cell_type2])].copy()\n",
    "            if specific_cells.shape[0] == 0:\n",
    "                specific_cells = adata3[adata3.obs['cell_type_scChat'].isin([standardized_cell_type3])].copy()\n",
    "            sc.tl.pca(specific_cells, svd_solver='arpack')\n",
    "            sc.pp.neighbors(specific_cells)\n",
    "            sc.tl.umap(specific_cells)\n",
    "            sc.tl.leiden(specific_cells, resolution=resolution)\n",
    "\n",
    "            specific_cells.obs['cell_type_scChat'] = 'Unknown'\n",
    "            for group, cell_type in map2.items():\n",
    "                specific_cells.obs.loc[specific_cells.obs['leiden'] == group, 'cell_type_scChat'] = cell_type\n",
    "            \n",
    "            # specific_cells.obs[['UMAP_1', 'UMAP_2', 'leiden', 'patient_name', 'cell_type_scChat']].to_csv(f'scChat/umaps/{standardized_cell_type}_annotated_umap_data.csv', index=False)\n",
    "            \n",
    "            annotated_adata = specific_cells.copy()\n",
    "            fname = f'annotated_adata/{standardized_cell_type}_annotated_adata.pkl'\n",
    "            with open(fname, \"wb\") as file:\n",
    "                pickle.dump(annotated_adata, file)\n",
    "\n",
    "    \n",
    "    except (SyntaxError, ValueError) as e:\n",
    "        print(f\"Error in parsing the map: {e}\")\n",
    "        # Handle the error or return an appropriate response\n",
    "\n",
    "    return \"Repeat 'Annotation of clusters is complete'\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Annotate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "# root = '/stor/lep/workspace/AISbenchmark/AISbench'\n",
    "root = '/stor/lep/workspace/AISbenchmark/Task1_data'\n",
    "# organ_list = os.listdir(root)\n",
    "organ_list = [file for file in os.listdir(root) if file.endswith('raw.h5ad')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------Bone_marrow_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/bone_marrow_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/bone_marrow_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/bone_marrow_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/bone_marrow_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/bone_marrow_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/bone_marrow_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/bone_marrow_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/bone_marrow_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/bone_marrow_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/bone_marrow_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/bone_marrow_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/bone_marrow_raw/normal.json\n",
      "-------------------Kidney_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/kidney_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/kidney_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/kidney_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/kidney_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/kidney_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/kidney_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/kidney_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/kidney_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/kidney_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/kidney_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/kidney_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/kidney_raw/normal.json\n",
      "-------------------Eye_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/eye_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/eye_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/eye_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/eye_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/eye_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/eye_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/eye_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/eye_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/eye_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/eye_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/eye_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/eye_raw/normal.json\n",
      "-------------------Blood_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/blood_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/blood_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/blood_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/blood_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/blood_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/blood_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/blood_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/blood_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/blood_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/blood_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/blood_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/blood_raw/normal.json\n",
      "-------------------Heart_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/heart_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/heart_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/heart_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/heart_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/heart_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/heart_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/heart_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/heart_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/heart_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/heart_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/heart_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/heart_raw/normal.json\n",
      "-------------------Trachea_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/trachea_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/trachea_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/trachea_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/trachea_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/trachea_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/trachea_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/trachea_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/trachea_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/trachea_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:394: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'names'] = self.var_names[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:396: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'scores'] = scores[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:399: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals'] = pvals[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:409: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'pvals_adj'] = pvals_adj[global_indices]\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_rank_genes_groups.py:420: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`\n",
      "  self.stats[group_name, 'logfoldchanges'] = np.log2(\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/trachea_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/trachea_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/trachea_raw/normal.json\n",
      "-------------------Testis_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/testis_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/testis_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/testis_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/testis_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/testis_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/testis_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/testis_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/testis_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/testis_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/testis_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/testis_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/testis_raw/normal.json\n",
      "-------------------Pancreas_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/pancreas_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/pancreas_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/pancreas_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/pancreas_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/pancreas_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/pancreas_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/pancreas_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/pancreas_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/pancreas_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/pancreas_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/pancreas_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/pancreas_raw/normal.json\n",
      "-------------------Adipose_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/adipose_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/adipose_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/adipose_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/adipose_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/adipose_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/adipose_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/adipose_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/adipose_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/adipose_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/adipose_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/adipose_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/adipose_raw/normal.json\n",
      "-------------------Liver_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/liver_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/liver_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/liver_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/liver_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/liver_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/liver_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/liver_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/liver_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/liver_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/liver_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/liver_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/liver_raw/normal.json\n",
      "-------------------Breast_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/breast_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/breast_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/breast_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/breast_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/breast_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/breast_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/breast_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/breast_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/breast_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/breast_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/breast_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/breast_raw/normal.json\n",
      "-------------------Salivary_gland_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/salivary_gland_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/salivary_gland_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/salivary_gland_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/salivary_gland_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/salivary_gland_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/salivary_gland_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/salivary_gland_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/salivary_gland_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/salivary_gland_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/salivary_gland_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/salivary_gland_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/salivary_gland_raw/normal.json\n",
      "-------------------Skin_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/skin_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/skin_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/skin_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/skin_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/skin_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/skin_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/skin_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/skin_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/skin_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/skin_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/skin_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/skin_raw/normal.json\n",
      "-------------------Intestine_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/intestine_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/intestine_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/intestine_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/intestine_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/intestine_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/intestine_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/intestine_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/intestine_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/intestine_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/intestine_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/intestine_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/intestine_raw/normal.json\n",
      "-------------------Ovary_raw.h5ad---------------------\n",
      "WARNING: You’re trying to run this on 3000 dimensions of `.X`, if you really want this, set `use_rep='X'`.\n",
      "         Falling back to preprocessing with `sc.pp.pca` and default params.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/anndata/_core/anndata.py:381: FutureWarning: The dtype argument is deprecated and will be removed in late 2024.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/ovary_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/ovary_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/ovary_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/ovary_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/ovary_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/ovary_raw/normal.json\n",
      "Constructed file paths: ['schatbot/scChat_RAG/human/ovary_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/ovary_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/ovary_raw/normal.json\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3088253/527890278.py:79: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = statistic_data.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/527890278.py:84: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = statistic_data.gt(0).groupby('leiden').mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/numpy/core/fromnumeric.py:86: FutureWarning: The behavior of DataFrame.sum with axis=None is deprecated, in a future version this will reduce over both axes and return a scalar. To retain the old behavior, pass axis=0 (or do not pass axis)\n",
      "  return reduction(axis=axis, out=out, **passkwargs)\n",
      "/home/lep/miniconda3/envs/aiscientist/lib/python3.10/site-packages/scanpy/tools/_dendrogram.py:133: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_df = rep_df.groupby(level=0).mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:137: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  mean_expression = marker_expression.groupby('leiden').mean()\n",
      "/tmp/ipykernel_3088253/3139913383.py:138: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.\n",
      "  expression_proportion = marker_expression.gt(0).groupby('leiden').mean()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Constructed file paths: ['schatbot/scChat_RAG/human/ovary_raw/normal.json']\n",
      "File not found: schatbot/scChat_RAG/human/ovary_raw/normal.json\n",
      "File not found: schatbot/scChat_RAG/human/ovary_raw/normal.json\n"
     ]
    }
   ],
   "source": [
    "overlap_organ = []\n",
    "for organ in organ_list:\n",
    "\n",
    "    # if organ.lower() not in os.listdir('scChat/schatbot/scChat_RAG/human'):\n",
    "    #     continue\n",
    "    print(f'-------------------{organ}---------------------')\n",
    "    overlap_organ.append(organ.split('.')[0])\n",
    "    # study = os.listdir(os.path.join(root,organ))[0]\n",
    "    # data_path = os.path.join(root,organ,study,'row_data.h5ad')\n",
    "    data_path = os.path.join(root,organ)\n",
    "    summary = generate_umap(data_path)\n",
    "\n",
    "    conversation_history = [\n",
    "        {'role': 'user', 'content': ' You are a chatbot for helping in Single Cell RNA Analysis, you can call the functions generate_umap, process_cells, label_clusters, display_umap, display_processed_umap and more multiple times. DO NOT FORGET THIS. respond with a greeting.'},\n",
    "        {'role': 'user', 'content': 'generate umap'},\n",
    "        {'role': 'user', 'content': summary},\n",
    "        {'role': 'user', 'content': 'annotate the cell type of each cluster based on the UMAP analysis results'}\n",
    "    ]\n",
    "    # break\n",
    "\n",
    "    new_response = openai.chat.completions.create(\n",
    "                        model=\"gpt-4.1\",\n",
    "                        messages=conversation_history,\n",
    "                        temperature=0.2,\n",
    "                        top_p=0.4\n",
    "                        # max_tokens=300\n",
    "                    )\n",
    "    last_message = new_response.choices[0].message.content\n",
    "    label_clusters('overall cells', last_message)\n",
    "    file_name = organ.replace('_raw','')\n",
    "    annotated_adata.write_h5ad(f'/stor/lep/workspace/AISbenchmark/scChat/annotated_adata/{file_name}')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aiscientist",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
